### PR TITLE
[BREAKING] Drop constant field from JSON AST

### PR DESCRIPTION
**Pull request by ekpyron**
Part of  #4090 that only removes ``constant`` from the JSON AST.

### CODE REVIEW

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 1]
```diff
@@ -325,8 +325,6 @@ bool ASTJsonConverter::visit(FunctionDefinition const& _node)
 	std::vector<pair<string, Json::Value>> attributes = {
 		make_pair("name", _node.name()),
 		make_pair("documentation", _node.documentation() ? Json::Value(*_node.documentation()) : Json::nullValue),
-		// FIXME: remove with next breaking release
-		make_pair(m_legacy ? "constant" : "isDeclaredConst", _node.stateMutability() <= StateMutability::View),
 		make_pair("payable", _node.isPayable()),
```
Comment (axic) [inline libsolidity/ast/ASTJsonConverter.cpp]: Should also remove `payable`.

File: Changelog.md [hunk 2]
```diff
@@ -32,6 +32,7 @@ Breaking Changes:
  * General: Remove assembly instruction aliases ``sha3`` and ``suicide``
  * General: C99-style scoping rules are enforced now. This was already the case in the experimental 0.5.0 mode.
  * General: Disallow combining hex numbers with unit denominations (e.g. ``0x1e wei``). This was already the case in the experimental 0.5.0 mode.
+ * JSON AST: Remove ``constant`` field.
```
Comment (axic) [inline Changelog.md]: Should mention that the same information is conveyed via `stateMutability`. (Should double check it is in the JSON)

File: test/libsolidity/StandardCompiler.cpp [hunk 3]
```diff
@@ -508,7 +508,7 @@ BOOST_AUTO_TEST_CASE(output_selection_dependent_contract_with_import)
 	Json::Value contract = getContractResult(result, "fileA", "A");
 	BOOST_CHECK(contract.isObject());
 	BOOST_CHECK(contract["abi"].isArray());
-	BOOST_CHECK_EQUAL(dev::jsonCompactPrint(contract["abi"]), "[{\"constant\":false,\"inputs\":[],\"name\":\"f\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"}]");
```
Comment (axic) [inline test/libsolidity/StandardCompiler.cpp]: This is the ABI so shouldn't be removed from there.

Comment (ekpyron) [reply test/libsolidity/StandardCompiler.cpp]: Ah, yes, right.

**Comment by ekpyron** [comment]
Also removes ``payable`` now.

File: test/libsolidity/ASTLegacyJSON.cpp [hunk 4]
```diff
@@ -266,16 +266,14 @@ BOOST_AUTO_TEST_CASE(function_type)
 	BOOST_CHECK_EQUAL(argument["attributes"]["name"], "x");
 	BOOST_CHECK_EQUAL(argument["attributes"]["type"], "function () payable external returns (uint256)");
 	Json::Value funType = argument["children"][0];
-	BOOST_CHECK_EQUAL(funType["attributes"]["constant"], false);
```
Comment (axic) [inline test/libsolidity/ASTLegacyJSON.cpp]: These test are for the legacy JSON (the `true` first parameter). Don't we have tests at all for the new one?!

Comment (ekpyron) [reply test/libsolidity/ASTLegacyJSON.cpp]: I was wondering the same - I was waiting for the tests to see, whether anything fails, but apparently it doesn't...

Comment (ekpyron) [reply test/libsolidity/ASTLegacyJSON.cpp]: There's a lot more tests in ``ASTLegacyJSON.cpp`` than in ``ASTJSON.cpp`` in general - maybe the idea was to only test cases for which the legacy flag makes a difference in ``ASTJSON.cpp`` and consider the rest covered by ``ASTLegacyJSON.cpp`` already?

That wouldn't have been true for ``constant`` though - but I'm not sure why the new JSON AST used an ``isDeclaredConst`` field there at all...

Comment (axic) [reply test/libsolidity/ASTLegacyJSON.cpp]: True, `ASTJSON` is for the new one and `ASTLegacyJSON` is for the old one. However it sems the legacy one has twice as many tests and this change is not covered in it.

Comment (ekpyron) [reply test/libsolidity/ASTLegacyJSON.cpp]: For a lot of cases the AST and the legacy AST are identical and I suspect we didn't want to duplicate the tests. I can move the tests that don't depend on the legacy flag from ``ASTLegacyJSON.cpp`` to ``ASTJSON.cpp`` (this way makes more sense IMHO) and verify that all cases that depend on the legacy flag have two versions and add test cases, if missing.

Comment (axic) [reply test/libsolidity/ASTLegacyJSON.cpp]: Sounds good.

**Comment by axic** [comment]
@soad003 is the Remix static analyzer using the old fields? I know it still uses the legacy ast.

File: test/libsolidity/ASTLegacyJSON.cpp (hunk 4 above)
Comment (ekpyron) [reply test/libsolidity/ASTLegacyJSON.cpp]: Or maybe I will duplicate all tests, just to be sure.

Comment (ekpyron) [reply test/libsolidity/ASTLegacyJSON.cpp]: Instead of improving the AST JSON tests here, it may be easier to consider merging #4677 first :)...

**Comment by axic** [comment]
It seems that Remix static analyser handles both old and new fields: https://github.com/ethereum/remix/blob/master/remix-analyzer/src/solidity-analyzer/modules/staticAnalysisCommon.js#L583-L604

Likely the breaking impact will be low.

File: test/libsolidity/ASTLegacyJSON.cpp (hunk 4 above)
Comment (ekpyron) [reply test/libsolidity/ASTLegacyJSON.cpp]: #4677 is merged, so now there's full testing.

File: test/libsolidity/ASTJSON/documentation.json [hunk 5]
```diff
@@ -148,7 +148,6 @@
 					"id" : 14,
 					"implemented" : true,
 					"isConstructor" : false,
```
Comment (axic) [inline test/libsolidity/ASTJSON/documentation.json:150]: Some other cleanup we could make: consistency between using the `is` prefix. Here we have `isConstructor` and `implemented`...

**Comment by soad003** [comment]
@axic yes I still use the old AST but as you pointed out i already handle the new stateMutability field. So if the constant field is not available anymore node.attributes.constant === true => false therefore the new stateMutability fields are used. So i guess from my side no changes immediate changes are needed.

File: test/libsolidity/ASTJSON/documentation.json (hunk 5 above)
Comment (axic) [reply test/libsolidity/ASTJSON/documentation.json:150]: Decision from the weekly meeting: only `isConstructor` uses the `is` prefix, lets remove it.

Also need to check if `fallback` is included.

Comment (ekpyron) [reply test/libsolidity/ASTJSON/documentation.json:150]: Now tracked in #4691 in 0.5.0 ``ready to be worked on``.

**Review by axic — APPROVED** [review]
Haven't verified the test output line-by-line though.

**Pull request merged**